core/src/main/java/org/javarosa/form/api/FormEntrySession.java

```diff
@@ -0,0 +1,214 @@
+    }
+
+    @Override
+    public void addQuestionSkip(FormIndex formIndex) {
+        final String formIndexString = formIndex.toString();
+        int insertIndex = removeDuplicateAction(formIndexString);
+        actions.insertElementAt(FormEntryAction.buildSkipAction(formIndexString), insertIndex);
+    }
+
+    public FormEntryAction popAction() {
+        if (actions.size() > 0) {
+            return actions.remove(0);
+        } else {
+            return FormEntryAction.buildNullAction();
+        }
+    }
+
+    public FormEntryAction peekAction() {
+        if (actions.size() > 0) {
+            return actions.get(0);
+        } else {
+            return FormEntryAction.buildNullAction();
+        }
+    }
+
+    public int size() {
+        return actions.size();
+    }
+
+    @Override
+    public String toString() {
+        StringBuilder sessionStringBuilder = new StringBuilder();
+
+        for (FormEntryAction formEntryAction : actions) {
+            sessionStringBuilder.append(formEntryAction).append(" ");
+        }
+
+        return sessionStringBuilder.toString().trim();
+    }
+
+    public static FormEntrySession fromString(String sessionString) {
+        FormEntrySession formEntrySession = new FormEntrySession();
+        for (String actionString : FormEntrySession.splitTopParens(sessionString)) {
+            formEntrySession.actions.addElement(FormEntryAction.fromString(actionString));
+        }
+
+        return formEntrySession;
+    }
+
+    public static class FormEntryAction implements Serializable {
+        public final String formIndexString;
+        public final String value;
+        public final String action;
+        private final static String NEW_REPEAT = "NEW_REPEAT";
+        private final static String SKIP = "SKIP";
+        private final static String VALUE = "VALUE";
+
+        private FormEntryAction(String formIndexString, String value, String action) {
+            this.formIndexString = formIndexString;
+            this.value = value;
+            this.action = action;
+        }
+
+        public static FormEntryAction buildNewRepeatAction(String formIndexString) {
+            return new FormEntryAction(formIndexString, "", NEW_REPEAT);
+        }
+
+        public static FormEntryAction buildValueSetAction(String formIndexString, String value) {
+            return new FormEntryAction(formIndexString, value, VALUE);
+        }
+
+        public static FormEntryAction buildSkipAction(String formIndexString) {
+            return new FormEntryAction(formIndexString, "", SKIP);
+        }
+
+        public static FormEntryAction buildNullAction() {
+            return new FormEntryAction("", "", "");
+        }
+
+        @Override
+        public String toString() {
+            if (NEW_REPEAT.equals(action)) {
+                return "((" + formIndexString + ") (NEW_REPEAT))";
+            } else if (VALUE.equals(action)) {
+                // TODO PLM: escape 'value' field
```

Comment:
Yeah, my hope for this was to work on it incrementally as it becomes more useful, since it is a side project of sorts. For instance, I don't think question lists are handled quite correctly yet, but I will fix that when I start needed to debug forms with question lists, or support does